keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,110 @@
+    }
+
+    /**
+     * Samples from the prior of a Probabilistic Model.
+     * Bayesian Network is the only supported implementation.
+     *
+     * Samples are taken by sampling from the prior of the desired variables in topological order.
+     *
+     * @param model the model to sample from
+     * @param variablesToSampleFrom the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        Variable latent = latentVariables.get(0);
+        Preconditions.checkArgument(latent instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex.");
+
+        BayesianNetwork network = new BayesianNetwork(((Vertex) latent).getConnectedGraph());
+
+        List<Vertex> observedVertices = network.getObservedVertices();
+        checkUpstreamOfObservedDoesNotContainProbabilistic(observedVertices);
+
+        List<Vertex> verticesToSampleFrom = variablesToSampleFrom.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+        Set<Vertex> allUpstreamVertices = allUpstreamVertices(verticesToSampleFrom);
+
+        List<Vertex> sortedVertices = TopologicalSort.sort(allUpstreamVertices);
+        sortedVertices = removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(sortedVertices);
+
+        return new ForwardSampler(verticesToSampleFrom, sortedVertices, random);
+    }
+
+    private List<Vertex> removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(List<Vertex> topologicalVertices) {
+        List<Vertex> copy = new ArrayList<>(topologicalVertices);
+
+        for (Vertex vertex : topologicalVertices) {
+            if (vertex instanceof Probabilistic) {
+                break;
+            } else {
+                copy.remove(vertex);
+            }
+        }
+        return copy;
+    }
+
+    private Set<Vertex> allUpstreamVertices(List<Vertex> fromVertices) {
+        Set<Vertex> upstream = new HashSet<>();
+        for (Vertex vertex : fromVertices) {
+            LambdaSection upstreamLambdaSection = LambdaSection.getUpstreamLambdaSection(vertex, true);
```

Comment:
This code is super clear - good job.  I'm a bit worried that this could cause pretty serious performance issues for larger graphs though (and yes - it only gets run once, but for interactive sampling that cost will still bite us).

I definitely think that this section provides a very strong incentive for enhancing the Lambda section code to improve performance for cases like this though.  Happy to help out with that if it would be useful.